### PR TITLE
Read Computer Vision facial landmarks and output classification using Emotion Synthesis format

### LINKED ISSUE #14: Be able to read in facial landmark positions as input from Computer Vision group
- [x] Parse Computer Vision input ([see format](https://gist.github.com/diegocasmo/9f5c025a0202e35f30d5f3c4af6242c7))
- [x] Export/import best model (either MLP or SVM)
- [x] Use best model predict Computer Vision input

Notes:
  - Kaylan mentioned in the meeting that we need to be able to tell what features are the most important ones and communicate that to the Computer Vision group. It would be a good idea to send an email to Kaylan to clarify this requirement.

### LINKED ISSUE #15: Classification results should be formatted such that they can be used by the Emotion Synthesis group
- [x] Create format for Emotion Synthesis group (must include accuracy and confidence)
  - [x] Update ``final-solution.ipynb`` notebook to include confusion matrix/accuracy
- [x] Send email to Emotion Synthesis group with format and a documentation of it
- [x] Classify a couple of examples and send them to the Emotion Synthesis (in the same email from above)

### PR DESCRIPTION
**Pull request by ankurshukla03**
Closes #14 
Closes #15 
``` python
{
  'ANGER': 0.00040305041178978045,
  'DISGUST': 0.2010905945567386,
  'FEAR': 0.0005442277116194078,
  'HAPPY': 0.7968281208888155,
  'SADNESS': 0.0011075836616630938,
  'SURPRISE': 2.6422769373673852e-05
}
```

I did not receive any mail from kalyan i am planning to send this format as an output to emotion synthesis group and it can be one of the example as its been generated using our MLP classifier model. Check solution notebook in emotion-synthesis branch for the steps.
- Do we need to send confusion matrix to the smotion synthesis group?(I think we do just to be sure)

### CODE REVIEW

**Comment by tristaaan** [comment]
@diegocasmo in the final method, we can just manually select the 15 components PCA deems important? Or do we have to run it still on the single row of data we get.

**Comment by diegocasmo** [comment]
> in the final method, we can just manually select the 15 components PCA deems important? Or do we have to run it still on the single row of data we get.

@tristaaan I'm not sure

**Comment by tristaaan** [comment]
Great find on the `pca.transform` I saved the `pca.fit(...` model and we can load it here. All runs well except the prediction.

I suspect (from experience in the other project) preprocessing is the culprit. Normalization specifically. 

```py
x = [[-3,-2,0,2,3], [-10, -8, 0, 8, 10]]
x = np.array(x)
normalize(x, axis=1)
#array([[-0.58834841, -0.39223227,  0.        ,  0.39223227,  0.58834841],
#       [-0.55215763, -0.4417261 ,  0.        ,  0.4417261 ,  0.55215763]])
normalize(x, axis=0)
#array([[-0.28734789, -0.24253563,  0.        ,  0.24253563,  0.28734789],
#       [-0.95782629, -0.9701425 ,  0.        ,  0.9701425 ,  0.95782629]])
```

`axis=1` `axis=0` makes quite the difference. We could normalize on the values we've been finding so far in the test set? Or recompute it all with `axis=1` (row by row normalization)

**Comment by tristaaan** [comment]
In the final notebook with `axis=1`  SVM retains its crown

```
kNN with PCA classifier accuracy: 	0.53%
SVM with PCA classifier accuracy: 	0.80%
MLP with PCA classifier accuracy: 	0.69%
Random Forest with PCA classifier accuracy: 	0.56%
```

**Comment by tristaaan** [comment]
Of all the silly coding standards out there one I really like is `import` ordering for python. 

1. System imports (`os`, `sys`, `json`...)
2. Third party imports (`numpy` ...)
3. Local imports (here `constants.py`)

Each category should be alphabetically ordered. It makes it a lot easier to read.

**Comment by ankurshukla03** [comment]
I am training SVM model with [CalibratedClassifierCV](http://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html) to use predict_proba function while giving output to emotion synthesis group as we need to give them the confidence with each of the labels.
Just to make sure can you guys check that the above function i am using is not wrong i read it and it seems fine to me, also it doesn't affect the accuracy of our SVM model.


```svm = CalibratedClassifierCV(clf)```

**Comment by tristaaan** [comment]
What email client are you using? That link goes to some malware for me,
click on it as few times as possible.

-Tristan Wright

**Comment by ankurshukla03** [comment]
I am using github itself, not any email client for that comment.

>  What email client are you using? That link goes to some malware for me,
click on it as few times as possible.

**Comment by tristaaan** [comment]
Lol I'm emailing all of these, that won't be an email client issue. Just check the link, markdown format will make it look nicer. `[cool site](http://zombo.com)`

**Comment by ankurshukla03** [comment]
Lol, Should have looked for the syntax first. Thanks btw.

> Just check the link, markdown format will make it look nicer.

**Comment by tristaaan** [comment]
The link itself though. It goes to the wrong scikit...
```
http://scikitlearn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html
vs
http://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html
```

**Comment by ankurshukla03** [comment]
Shitt. Changed the link now. 
> The link itself though. It goes to the wrong scikit...

**Comment by tristaaan** [comment]
Is the PCA model different too since we're normalizing differently?

**Comment by ankurshukla03** [comment]
Oops. Changing it now
> Is the PCA model different too since we're normalizing differently?

**Comment by tristaaan** [comment]
Cool, I think if the meeting tomorrow goes ok we can make the input changes they're feeding us merge this.

**Comment by tristaaan** [comment]
Models generated by exporting the data-analysis notebook to a script and running it from the command line with python 2 explicitly.

**Comment by tristaaan** [comment]
Pretty certain this can be merged now.

**Comment by diegocasmo** [comment]
Let's wait until tomorrow, minor changes may occur during the meeting. The downgrade of all the notebooks to Python 2 can be done in a different PR.

**Comment by tristaaan** [comment]
I have no idea what to do with this branch now. Merge away after conflicts resolved?

**Comment by diegocasmo** [comment]
@tristaaan Don't you need the changes in this branch for the presentation?

**Comment by tristaaan** [comment]
That's what it's for. 
Yes I resolved the merge conflict. Merging away.
